File: scripts/large_files.lua
#!/usr/bin/env lua
-- Find and display the largest files on the filesystem with deletion suggestions.

-- ── categories ──────────────────────────────────────────────────────────────

local EXT_TO_CAT = {}
local function def(cat, ...)
  for _, e in ipairs({...}) do EXT_TO_CAT["." .. e] = cat end
end
def("Video",      "mp4","mkv","avi","mov","wmv","flv","webm","m4v","ts","vob")
def("Audio",      "mp3","flac","wav","aac","ogg","m4a","wma","opus")
def("Image",      "jpg","jpeg","png","gif","bmp","tiff","webp","heic","raw","cr2","nef")
def("Archive",    "zip","tar","gz","bz2","xz","7z","rar","zst","tgz","tbz2","tbz")
def("Disk Image", "iso","img","vmdk","vdi","qcow2","dmg")
def("Document",   "pdf","doc","docx","xls","xlsx","ppt","pptx","odt","ods")
def("Code",       "py","js","ts","go","rs","c","cpp","h","java","rb","sh")
def("Database",   "db","sqlite","sqlite3","sql")
def("Package",    "deb","rpm","pkg","apk","msi","exe")
def("Log",        "log")
def("Cache",      "cache")
def("Backup",     "bak","old","orig","backup")
def("Library",    "so","dylib","dll","a")

local DEL_PATHS = {
  "/.cache/","/tmp/","/var/tmp/","/var/cache/","/.local/share/Trash/",
  "/__pycache__/","/node_modules/","/.npm/_cacache/","/.cargo/registry/cache/",
  "/go/pkg/mod/cache/","/.gradle/caches/","/.m2/repository/",
  "/.thumbnails/","/thumbnails/",
}
local DEL_EXTS  = { [".log"]=true,[".bak"]=true,[".old"]=true,[".orig"]=true,
                    [".backup"]=true,[".cache"]=true,[".pyc"]=true,[".pyo"]=true }
local DEL_NAMES = { ["core"]=true,["core.gz"]=true,[".DS_Store"]=true,
                    ["Thumbs.db"]=true,["desktop.ini"]=true,
                    ["npm-debug.log"]=true,["yarn-error.log"]=true }

-- ── helpers ──────────────────────────────────────────────────────────────────

local function categorize(path, ext)
  local p = path:lower()
  if p:find("/node_modules/", 1, true) then return "Package" end
  if p:find("/__pycache__/",  1, true) or ext == ".pyc" then return "Cache" end
  if p:find("/.cache/", 1, true) or p:find("/cache/", 1, true) then return "Cache" end
  if p:find("/log/",  1, true) or p:find("/logs/", 1, true) then return "Log" end
  return EXT_TO_CAT[ext:lower()] or "Other"
end

local function is_deletable(path, ext, name)
  local p = path:lower()
  for _, pat in ipairs(DEL_PATHS) do
    if p:find(pat, 1, true) then
      return true, "in " .. pat:gsub("^/",""):gsub("/$","")
    end
  end
  if DEL_EXTS[ext:lower()]   then return true, ext .. " file"   end
  if DEL_NAMES[name:lower()] then return true, "temp/junk file" end
  return false, ""
end

local function human_size(n)
  for _, u in ipairs({"B","KB","MB","GB","TB"}) do
    if n < 1024 then return ("%.1f %s"):format(n, u) end
    n = n / 1024
  end
  return ("%.1f PB"):format(n)
end

local function shell_quote(s)
  return "'" .. s:gsub("'", "'\\''") .. "'"
end

local function term_width()
  local h = io.popen("tput cols 2>/dev/null")
  local w = h and tonumber(h:read("*l"))
  if h then h:close() end
  return math.min(w or 120, 160)
end

-- ── scan ─────────────────────────────────────────────────────────────────────

local function scan(roots, skip_mounts, min_size)
  local seen    = {}
  local entries = {}

  for _, root in ipairs(roots) do
    local xdev = skip_mounts and "-xdev" or ""
    -- find prints null-terminated paths; xargs batches them into stat calls.
    -- stat '%d %i %z %N' → "dev ino size /full/path" (path always last, may have spaces)
    local cmd = ("find %s %s -type f -size +%dc -print0 2>/dev/null"
              .. " | xargs -0 stat -f '%%d %%i %%z %%N' 2>/dev/null"):format(
                   shell_quote(root), xdev, min_size - 1)

    local h = io.popen(cmd)
    if not h then
      io.stderr:write("error: cannot scan " .. root .. "\n")
    else
      for line in h:lines() do
        local dev, ino, sz, path = line:match("^(%d+) (%d+) (%d+) (.+)$")
        if dev then
          local key = dev .. ":" .. ino
          if not seen[key] then
            seen[key] = true
            local name = path:match("[^/]+$") or path
            local ext  = name:match("(%.[^.]+)$") or ""
            local cat  = categorize(path, ext)
            local del, reason = is_deletable(path, ext, name)
            entries[#entries + 1] = {
              path = path, size = tonumber(sz),
              category = cat, deletable = del, reason = reason,
            }
          end
        end
      end
      h:close()
    end
  end

  table.sort(entries, function(a, b) return a.size > b.size end)
  return entries
end

-- ── output ───────────────────────────────────────────────────────────────────

local function print_table(entries, top_n, color)
  local n   = math.min(top_n, #entries)
  local w   = term_width()
  local sep = ("-"):rep(w)

  local RED  = color and "\27[31m"   or ""
  local HEAD = color and "\27[1;36m" or ""
  local RST  = color and "\27[0m"    or ""

  print(("\n%sTop %d Largest Files%s"):format(HEAD, n, RST))
  print(sep)
  print(("%4s  %10s  %-12s  %-5s  %-22s  Path"):format(
        "#", "Size", "Category", "Del?", "Reason"))
  print(sep)

  local total, del_size, del_count = 0, 0, 0
  for i = 1, n do
    local e   = entries[i]
    local del = e.deletable and "  *  " or "     "
    local row = ("%4d  %10s  %-12s  %-5s  %-22s  %s"):format(
                  i, human_size(e.size), e.category, del, e.reason, e.path)
    if color and e.deletable then
      print(RED .. row .. RST)
    else
      print(row)
    end
    total = total + e.size
    if e.deletable then
      del_size  = del_size  + e.size
      del_count = del_count + 1
    end
  end

  print(sep)
  print(("Total shown: %s  |  Flagged deletable: %d files (%s)"):format(
        human_size(total), del_count, human_size(del_size)))
end

-- ── arg parsing ──────────────────────────────────────────────────────────────

local function parse_args()
  local opts = { roots={}, top=100, min_size=1024*1024, skip_mounts=true, plain=false }
  local i = 1
  while i <= #arg do
    local a = arg[i]
    if a == "-h" or a == "--help" then
      io.write(
        "usage: large_files.lua [opts] [roots...]\n"
        .. "  -n N, --top N     files to show (default: 100)\n"
        .. "  --min-size BYTES  minimum size in bytes (default: 1048576)\n"
        .. "  --no-skip-mounts  cross filesystem boundaries\n"
        .. "  --plain           disable ANSI color\n"
        .. "  roots             paths to scan (default: /)\n")
      os.exit(0)
    elseif a == "-n" or a == "--top" then
      i = i + 1; opts.top = tonumber(arg[i]) or opts.top
    elseif a:match("^%-n(%d+)$") then
      opts.top = tonumber(a:match("^%-n(%d+)$"))
    elseif a == "--min-size" then
      i = i + 1; opts.min_size = tonumber(arg[i]) or opts.min_size
    elseif a == "--no-skip-mounts" then
      opts.skip_mounts = false
    elseif a == "--plain" then
      opts.plain = true
    elseif not a:match("^%-") then
      opts.roots[#opts.roots + 1] = a
    else
      io.stderr:write("unknown option: " .. a .. "\n"); os.exit(1)
    end
    i = i + 1
  end
  if #opts.roots == 0 then opts.roots = {"/"} end
  return opts
end

-- ── main ─────────────────────────────────────────────────────────────────────

local opts  = parse_args()
local color = not opts.plain and os.execute("[ -t 1 ]") == true

io.stderr:write(("Scanning %s (min size: %s) ...\n"):format(
  table.concat(opts.roots, ", "), human_size(opts.min_size)))

local entries = scan(opts.roots, opts.skip_mounts, opts.min_size)

io.stderr:write(("Found %d files above threshold.\n"):format(#entries))

if #entries == 0 then
  print("No files found.")
  os.exit(0)
end

print_table(entries, opts.top, color)

File: scripts/large_files.py
#!/usr/bin/env python3
"""Find and display the largest files on the filesystem with deletion suggestions."""

import os
import sys
import stat
import argparse
from pathlib import Path
from dataclasses import dataclass, field
from typing import Optional
import shutil

# Optional rich for pretty output
try:
    from rich.console import Console
    from rich.table import Table
    from rich import box
    HAS_RICH = True
except ImportError:
    HAS_RICH = False


CATEGORIES = {
    "Video":     {".mp4", ".mkv", ".avi", ".mov", ".wmv", ".flv", ".webm", ".m4v", ".ts", ".vob"},
    "Audio":     {".mp3", ".flac", ".wav", ".aac", ".ogg", ".m4a", ".wma", ".opus"},
    "Image":     {".jpg", ".jpeg", ".png", ".gif", ".bmp", ".tiff", ".webp", ".heic", ".raw", ".cr2", ".nef"},
    "Archive":   {".zip", ".tar", ".gz", ".bz2", ".xz", ".7z", ".rar", ".zst", ".tgz", ".tbz2", ".tbz"},
    "Disk Image":{".iso", ".img", ".vmdk", ".vdi", ".qcow2", ".dmg"},
    "Document":  {".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx", ".odt", ".ods"},
    "Code":      {".py", ".js", ".ts", ".go", ".rs", ".c", ".cpp", ".h", ".java", ".rb", ".sh"},
    "Database":  {".db", ".sqlite", ".sqlite3", ".sql"},
    "Package":   {".deb", ".rpm", ".pkg", ".apk", ".msi", ".exe"},
    "Container": {".tar"},  # docker save output — also matched by Archive, but we check suffix combos below
    "Log":       {".log"},
    "Cache":     {".cache"},
    "Backup":    {".bak", ".old", ".orig", ".backup"},
}

# Paths whose contents are usually safe to delete
DELETABLE_PATH_PATTERNS = [
    "/.cache/",
    "/tmp/",
    "/var/tmp/",
    "/var/cache/",
    "/.local/share/Trash/",
    "/__pycache__/",
    "/node_modules/",
    "/.npm/_cacache/",
    "/.cargo/registry/cache/",
    "/go/pkg/mod/cache/",
    "/.gradle/caches/",
    "/.m2/repository/",
    "/.thumbnails/",
    "/thumbnails/",
]

DELETABLE_EXTENSIONS = {".log", ".bak", ".old", ".orig", ".backup", ".cache", ".pyc", ".pyo"}

DELETABLE_NAMES = {
    "core", "core.gz", ".DS_Store", "Thumbs.db", "desktop.ini",
    "npm-debug.log", "yarn-error.log",
}


def categorize(path: str, ext: str) -> str:
    ext_lower = ext.lower()
    for cat, exts in CATEGORIES.items():
        if ext_lower in exts:
            return cat
    # Guess from path hints
    p = path.lower()
    if "/node_modules/" in p:
        return "Package"
    if "/__pycache__/" in p or ".pyc" in p:
        return "Cache"
    if "/.cache/" in p or "/cache/" in p:
        return "Cache"
    if "/log/" in p or "/logs/" in p:
        return "Log"
    if ext_lower in {".so", ".dylib", ".dll", ".a"}:
        return "Library"
    if ext_lower in {".bin", ""}:
        return "Binary"
    return "Other"


def is_deletable(path: str, ext: str, name: str) -> tuple[bool, str]:
    p = path.lower()
    for pattern in DELETABLE_PATH_PATTERNS:
        if pattern.lower() in p:
            return True, f"in {pattern.strip('/')}"
    if ext.lower() in DELETABLE_EXTENSIONS:
        return True, f"{ext} file"
    if name.lower() in DELETABLE_NAMES:
        return True, "temp/junk file"
    return False, ""


@dataclass
class FileEntry:
    path: str
    size: int
    category: str
    deletable: bool
    delete_reason: str


def human_size(n: int) -> str:
    for unit in ("B", "KB", "MB", "GB", "TB"):
        if n < 1024:
            return f"{n:.1f} {unit}"
        n /= 1024
    return f"{n:.1f} PB"


def scan(roots: list[str], skip_mounts: bool, min_size: int) -> list[FileEntry]:
    seen_inodes: set[int] = set()
    entries: list[FileEntry] = []
    root_dev = {r: os.stat(r).st_dev for r in roots}

    def should_skip(path: str, dev: int) -> bool:
        if skip_mounts:
            try:
                if os.stat(path).st_dev != dev:
                    return True
            except OSError:
                return True
        return False

    for root in roots:
        dev = root_dev[root]
        for dirpath, dirnames, filenames in os.walk(root, followlinks=False, onerror=lambda e: None):
            # Prune mount points when requested
            if skip_mounts:
                try:
                    dir_dev = os.stat(dirpath).st_dev
                except OSError:
                    dirnames.clear()
                    continue
                if dir_dev != dev:
                    dirnames.clear()
                    continue
            # Skip proc/sys early
            if dirpath in ("/proc", "/sys", "/dev"):
                dirnames.clear()
                continue

            for fname in filenames:
                fpath = os.path.join(dirpath, fname)
                try:
                    st = os.lstat(fpath)
                except OSError:
                    continue
                if not stat.S_ISREG(st.st_mode):
                    continue
                if st.st_size < min_size:
                    continue
                # Skip hard-link duplicates
                inode_key = (st.st_dev, st.st_ino)
                if inode_key in seen_inodes:
                    continue
                seen_inodes.add(inode_key)

                ext = Path(fname).suffix
                cat = categorize(fpath, ext)
                deletable, reason = is_deletable(fpath, ext, fname)
                entries.append(FileEntry(fpath, st.st_size, cat, deletable, reason))

    entries.sort(key=lambda e: e.size, reverse=True)
    return entries


def print_table_rich(entries: list[FileEntry], top_n: int) -> None:
    console = Console()
    table = Table(
        title=f"Top {min(top_n, len(entries))} Largest Files",
        box=box.ROUNDED,
        show_lines=False,
        header_style="bold cyan",
    )
    table.add_column("#", justify="right", style="dim", width=4)
    table.add_column("Size", justify="right", width=10)
    table.add_column("Category", width=12)
    table.add_column("Del?", justify="center", width=5)
    table.add_column("Reason", width=22)
    table.add_column("Path")

    for i, e in enumerate(entries[:top_n], 1):
        del_mark = "[red]✗[/red]" if e.deletable else ""
        row_style = "red" if e.deletable else ""
        table.add_row(
            str(i),
            human_size(e.size),
            e.category,
            del_mark,
            e.delete_reason,
            e.path,
            style=row_style,
        )
    console.print(table)

    total = sum(e.size for e in entries[:top_n])
    deletable = [e for e in entries[:top_n] if e.deletable]
    deletable_size = sum(e.size for e in deletable)
    console.print(f"\n[bold]Total shown:[/bold] {human_size(total)}  "
                  f"[bold]Flagged deletable:[/bold] [red]{len(deletable)} files "
                  f"({human_size(deletable_size)})[/red]")


def print_table_plain(entries: list[FileEntry], top_n: int) -> None:
    col_widths = (4, 10, 12, 5, 22)
    header = f"{'#':>4}  {'Size':>10}  {'Category':<12}  {'Del?':^5}  {'Reason':<22}  Path"
    sep = "-" * min(shutil.get_terminal_size().columns, 120)
    print(f"\nTop {min(top_n, len(entries))} Largest Files")
    print(sep)
    print(header)
    print(sep)
    for i, e in enumerate(entries[:top_n], 1):
        del_mark = "  *  " if e.deletable else "     "
        print(f"{i:>4}  {human_size(e.size):>10}  {e.category:<12}  {del_mark}  {e.delete_reason:<22}  {e.path}")
    print(sep)
    total = sum(e.size for e in entries[:top_n])
    deletable_size = sum(e.size for e in entries[:top_n] if e.deletable)
    n_del = sum(1 for e in entries[:top_n] if e.deletable)
    print(f"Total shown: {human_size(total)}  |  Flagged deletable: {n_del} files ({human_size(deletable_size)})")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Find the largest files on the filesystem and flag ones safe to delete."
    )
    parser.add_argument(
        "roots", nargs="*", default=["/"],
        help="Root paths to scan (default: /)"
    )
    parser.add_argument(
        "-n", "--top", type=int, default=100,
        help="Number of files to show (default: 100)"
    )
    parser.add_argument(
        "--min-size", type=int, default=1024 * 1024,
        help="Minimum file size in bytes to consider (default: 1 MB)"
    )
    parser.add_argument(
        "--no-skip-mounts", action="store_true",
        help="Cross filesystem boundaries (default: stay on the same device per root)"
    )
    parser.add_argument(
        "--plain", action="store_true",
        help="Plain text output (no rich formatting)"
    )
    args = parser.parse_args()

    skip_mounts = not args.no_skip_mounts

    print(f"Scanning {', '.join(args.roots)} (min size: {human_size(args.min_size)}) ...", file=sys.stderr)
    entries = scan(args.roots, skip_mounts, args.min_size)
    print(f"Found {len(entries):,} files above threshold.", file=sys.stderr)

    if not entries:
        print("No files found.")
        return

    if HAS_RICH and not args.plain:
        print_table_rich(entries, args.top)
    else:
        print_table_plain(entries, args.top)


if __name__ == "__main__":
    main()
